src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.Helpers.cs

```diff
@@ -0,0 +1,274 @@
+                // Check if the invocation of the deferred execution method is used in another deferred execution method.
+                return SkipDeferredAndConversionMethodIfNeeded(operation.Parent.Parent, wellKnownSymbolsInfo);
+            }
+
+            return operation;
+        }
+
+        private static bool IsImplicitConversion(IOperation operation)
+        {
+            // Check if this is a conversion operation targeting IEnumerable<T> or IEnumerable,
+            // and this conversion is passed into another deferred execution method.
+            // This is used in methods chain like
+            // 1. Cast<T> and OfType<T>, which takes IEnumerable as the first parameter. For example:
+            // c.Select(i => i + 1).Cast<long>();
+            // 'c.Select(i => i + 1)' has IEnumerable<T> type, and will be implicitly converted to IEnumerable. Then the conversion result would be passed to Cast<long>().
+            // 2. OrderBy, ThenBy, etc.. which returns IOrderedIEnumerable<T>. For this example,
+            // c.OrderBy(i => i.Key).Select(m => m + 1);
+            // 'c.OrderBy(i => i.Key)' has IOrderedIEnumerable<T> type, and will be implicitly converted to IEnumerable<T> . Then the conversion result would be passed to Select()
+            return operation is IConversionOperation { IsImplicit: true } conversionOperation
+                   && conversionOperation.Type.OriginalDefinition.SpecialType is SpecialType.System_Collections_Generic_IEnumerable_T or SpecialType.System_Collections_IEnumerable;
+        }
+
+        /// <summary>
+        /// Check if the LocalReferenceOperation or ParameterReferenceOperation is enumerated by for each loop
+        /// </summary>
+        private static bool IsOperationEnumeratedByForEachLoop(
+            IOperation operation,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo)
+        {
+            RoslynDebug.Assert(operation is ILocalReferenceOperation or IParameterReferenceOperation);
+            if (!IsDeferredType(operation.Type.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                return false;
+            }
+
+            var operationToCheck = SkipDeferredAndConversionMethodIfNeeded(operation, wellKnownSymbolsInfo);
+            return operationToCheck.Parent is IForEachLoopOperation;
+        }
+
+        private static bool IsOperationEnumeratedByInvocation(
+            IOperation operation,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo)
+        {
+            if (operation.Parent is IArgumentOperation { Parent: IInvocationOperation invocationOperation } argumentOperation)
+            {
+                return IsInvocationCausingEnumerationOverArgument(
+                    invocationOperation,
+                    argumentOperation,
+                    wellKnownSymbolsInfo);
+            }
+
+            return false;
+        }
+
+        private static bool IsInvocationCausingEnumerationOverArgument(
```

Comment:
Can we add a doc comment with an example? You seem to have detailed comments with examples in methods above, so doing so for this method as well would be good.